crypto/ess/ess_lib.c

```diff
@@ -290,72 +290,83 @@ static int ess_issuer_serial_cmp(const ESS_ISSUER_SERIAL *is, const X509 *cert)
+        /* TODO(3.0): fetch sha algorithm from providers */
+        if (cid != NULL)
+            md = EVP_sha1();
+        else
+            md = cid_v2->hash_alg == NULL ? EVP_sha256() :
+                EVP_get_digestbyobj(cid_v2->hash_alg->algorithm);
+        if (md == NULL) {
+            ERR_raise(ERR_LIB_ESS, ESS_R_ESS_DIGEST_ALG_UNKNOWN);
+            return -1;
+        }
 
-        if (cid->hash->length == SHA_DIGEST_LENGTH
-            && memcmp(cid->hash->data, cert_sha1, SHA_DIGEST_LENGTH) == 0) {
-            const ESS_ISSUER_SERIAL *is = cid->issuer_serial;
+        cid_hash_len = cid != NULL ? cid->hash->length : cid_v2->hash->length;
+        if (!X509_digest(cert, md, cert_digest, &len)
+                || cid_hash_len != len) {
+            ERR_raise(ERR_LIB_ESS, ESS_R_ESS_CERT_DIGEST_ERROR);
+            return -1;
+        }
 
+        if (memcmp(cid != NULL ? cid->hash->data : cid_v2->hash->data,
+                   cert_digest, len) == 0) {
+            is = cid != NULL ? cid->issuer_serial : cid_v2->issuer_serial;
             if (is == NULL || ess_issuer_serial_cmp(is, cert) == 0)
                 return i;
         }
     }
 
+    ERR_raise(ERR_LIB_ESS, ESS_R_ESS_CERT_ID_NOT_FOUND);
     return -1;
 }
 
-/* Returns < 0 if certificate is not found, certificate index otherwise. */
-int ossl_ess_find_cert_v2(const STACK_OF(ESS_CERT_ID_V2) *cert_ids,
-                          const X509 *cert)
+/*
+ * Check if first ESSCertID{,v2} matches signer cert (first in chain)
+ * and each further ESSCertID{,v2} matches any furher cert in the chain.
+ */
+int ossl_ess_check_signing_certs(const ESS_SIGNING_CERT *ss,
+                                 const ESS_SIGNING_CERT_V2 *ssv2,
+                                 const STACK_OF(X509) *chain)
 {
-    int i;
-    unsigned char cert_digest[EVP_MAX_MD_SIZE];
-    unsigned int len;
-
-    /* Look for cert in the cert_ids vector. */
-    for (i = 0; i < sk_ESS_CERT_ID_V2_num(cert_ids); ++i) {
-        const ESS_CERT_ID_V2 *cid = sk_ESS_CERT_ID_V2_value(cert_ids, i);
-        const EVP_MD *md;
-
-        if (cid == NULL)
-            return -1;
-        if (cid->hash_alg != NULL)
-            md = EVP_get_digestbyobj(cid->hash_alg->algorithm);
-        else
-            md = EVP_sha256();
-
-        /* TODO(3.0): fetch sha1 algorithm from providers */
-        if (!X509_digest(cert, md, cert_digest, &len))
-            return -1;
-
-        if (cid->hash->length != (int)len)
-            return -1;
-
-        if (memcmp(cid->hash->data, cert_digest, cid->hash->length) == 0) {
-            const ESS_ISSUER_SERIAL *is = cid->issuer_serial;
-
-            if (is == NULL || ess_issuer_serial_cmp(is, cert) == 0)
-                return i;
+    int i, j;
+
+    if (ss != NULL)
+        for (i = 0; i < sk_ESS_CERT_ID_num(ss->cert_ids); i++) {
+            j = find_cid(sk_ESS_CERT_ID_value(ss->cert_ids, i), NULL, chain);
+            if (j < 0)
```

Comment:
Nothing raise here?  
This puzzled me a bit; I am not familiar with the codebase. If the explanation can be short, could you explain how this `ERR_raise` thing works? I think of `throw` in Java but I didn't know c has a similar practice.

Reply:
> Nothing raise here?

Good point - the idea was that `find_cid()` should have raised an error here,
which was true for most but not all cases.
I've just made sure that also in all error situations an error is thrown by `find_cid()`.

Reply:
> If the explanation can be short, could you explain how this `ERR_raise` thing works?
> I think of throw in Java but I didn't know c has a similar practice.

This pushes an entry on the error queue,
but in contrast to Java the control flow must be done manually in C.